pkg/router/template/template_helper.go

```diff
@@ -386,6 +387,27 @@ func parseIPList(list string) string {
 	return list
 }
 
+// maxTimeoutFirstMatchedAndClipped finds the maximum timeout managed by a given annotation among all the routes, matches it against a given pattern, and clips.
+func maxTimeoutFirstMatchedAndClipped(aliases map[ServiceAliasConfigKey]ServiceAliasConfig, annotation, pattern string, values ...string) string {
+	var (
+		max         string
+		maxDuration time.Duration
+	)
+	for _, cfg := range aliases {
+		timeout := clipHAProxyTimeoutValue(firstMatch(pattern, append([]string{cfg.Annotations[annotation]}, values...)...))
+		if timeout != "" {
+			// No error handling because clipHAProxyTimeoutValue returns
+			// a valid timeout or an empty string. The latter is already handled.
+			timeoutDuration, _ := time.ParseDuration(timeout)
```

Comment:
This is a lot of redundant computation:

* `maxTimeoutFirstMatchedAndClipped` parses each route's timeout annotation value using `firstMatch`, then (if the value is valid) `clipHAProxyTimeoutValue`, then `time.ParseDuration`.
* `fe_sni` calls `maxTimeoutFirstMatchedAndClipped`, so you're parsing each valid value 3 times.
* `fe_no_sni` also calls `maxTimeoutFirstMatchedAndClipped`, so you're parsing each valid value 3 more times.
* The backend calls `clipHAProxyTimeoutValue (firstMatch $timeSpecPattern ...)`, so you're parsing each valid value 2 more times.

Granted, the regexp cache that Andy added in #268 avoids the cost of compiling the same regexp repeatedly, but you're still repeatedly doing a regexp match and then `time.ParseDuration` on the same value, right?

Reply:
The goal of the function was to define which timeout value will be set on the lowest layer backends. That's why I used exactly the same combination of function which is used in the template.

After some time spent on this one I didn't find a lot of room for the improvement here. I implemented a benchmark test which:
- allowed me to find a faster way to do the same: the max annotation timeout is now found first and the values are computed only once afterwards.
- showed that I don't get any visible performance gain if I try to get the `time.Duration` from `clipHAProxyTimeoutValue` to avoid another call to `ParseDuration`.

Also, [as you advised](https://github.com/openshift/router/pull/536#discussion_r1488665727) I moved the calculation of the max server timeout out of the backends so that it's done only one.